waiter/src/waiter/metrics_consumer.clj

```diff
@@ -0,0 +1,193 @@
+                      (doseq [event object]
+                        (async/put! event-chan event)))
+                    (log/warn "received unknown metrics event" {:type type})))
+                (log/warn "watch request was closed unexpectedly" {:error-chan (when error-chan (async/<! error-chan))
+                                                                   :metrics-endpoint metrics-endpoint}))
+              (catch Exception e
+                (log/error e "watch request to external request metrics service failed. Going to start retrying"
+                           {:url url})))
+            ; unnecessary to do exponential backoff because there is a static number of clients (waiter routers) making watch requests
+            (log/info (str "waiting " retry-delay-ms " ms before attempting to make the watch connection"))
+            (let [timeout-ch (async/timeout retry-delay-ms)
+                  [msg current-chan]
+                  (async/alts! [exit-chan timeout-ch] :priority true)]
+              (condp = current-chan
+                exit-chan (log/warn "exiting watch because exit chan was triggered" {:msg msg})
+                timeout-ch (recur))))]
+      {:exit-chan exit-chan
+       :go-chan go-chan
+       :query-state-fn query-state-fn})))
+
+(defn start-metrics-consumer-maintainer
+  "Initializes all external metrics service watch connections and filters/processes metric events that are then exposed
+  for other components to listen on to with `async/tap`.
+
+  Returns a map with keys:
+  :exit-chan is as channel that when message is pushed to it, causes all watch connections to be terminated
+  :query-state-fn is a query function for introspecting the state of the metrics events channels and watches
+  :token-metric-chan-mult is a mult for other components to listen in on when last-request-time was updated for a token"
+  [http-client clock kv-store token-cluster-calculator retrieve-descriptor-fn service-id->metrics-fn
+   make-metrics-watch-request-fn local-usage-agent router-id metrics-services token-metric-chan-buffer-size
+   retry-delay-ms]
+  (cid/with-correlation-id
+    "metrics-consumer-maintainer"
+    (let [exit-chan (async/promise-chan)
+          exit-mult (async/mult exit-chan)
+          correlation-id (cid/get-correlation-id)
+          default-cluster (token/get-default-cluster token-cluster-calculator)
+          state-atom (atom {:last-token-event-time nil})
+          token-metric-chan-buffer (async/buffer token-metric-chan-buffer-size)
+          token-metric-chan
+          (async/chan
+            token-metric-chan-buffer
+            (comp
+              (map (fn format-event
+                     [raw-event]
+                     (let [event (walk/keywordize-keys raw-event)
+                           kabob-event (set/rename-keys event {:lastRequestTime :last-request-time})]
+                       (update kabob-event :last-request-time f/parse))))
```

Comment:
Nitpick: this feels wasteful. We can `dissoc` the bad keys and `assoc` the formatted value.